_posts/2024-06-14-granite-rag.adoc

```diff
@@ -0,0 +1,286 @@
+
+With Quarkus, implementing an AI service is straightforward:
+
+[source,java]
+----
+@RegisterAiService(retrievalAugmentor = Retriever.class)
+@SystemMessage("""
+    You are Mona, a chatbot answering question about a museum. Be polite, concise and helpful.
+""")
+@SessionScoped
+public interface ChatBot {
+
+    String chat(String question);
+
+}
+----
+
+The `@RegisterAiService` annotation specifies the retrieval augmentor to use, which in our case is the `Retriever` bean defined earlier.
+
+The `@SystemMessage` annotation provides the main instructions for the AI model.
+
+The `@SessionScoped` annotation ensures that the AI service is stateful, maintaining context between user interactions for more engaging conversations.
+
+The `ChatBot` interface defines a single method, `chat`, which takes a user question as input and returns the chatbot's response.
+
+
+
+=== The WebSocket endpoint
+
+The final piece is the https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/ChatWebSocket.java[WebSocket endpoint], which serves as the communication bridge between the chatbot's backend and the frontend interface:
+
+[source,java]
+----
+@WebSocket(path = "/chat")
+public class ChatWebSocket {
+
+    @Inject ChatBot bot; // Inject the AI service
+
+    @OnOpen
+    String welcome() {
+        return "Welcome, my name is Mona, how can I help you today?";
+    }
+
+    @OnTextMessage
+    String onMessage(String message) {
+        return bot.chat(message);
+    }
+
+}
+----
+
+The `@WebSocket` annotation specifies the WebSocket path.
+The `@OnOpen` method sends a welcome message when a user connects to the _WebSocket_.
+The `@OnTextMessage` method processes the user's messages and returns the chatbot's responses, using the injected AI service.
+
+That's it! Our chatbot is now ready to chat with users, providing contextually relevant responses based on the RAG pattern.
+
+== Running the application
+
+Let's run the application and see our chatbot in action.
+First, clone the https://github.com/cescoffier/quarkus-granite-rag-demo/tree/main[repository] and run the following command:
+
+[source,shell]
+----
+mvn quarkus:dev
+----
+
+This command starts the Quarkus application in development mode.
+Ensure you have Instructlab or Podman AI Studio running to use the Granite LLM.
+You will also need Docker or Podman to automatically start Infinispan.
+
+Once the application is up and running, open your browser and navigate to http://localhost:8080.
+You should see the chatbot interface, where you can start chatting with Mona:
+
+image::mona-screenshot.png[The Mon chatbot,float="right",align="center"]
+
+== Summary
+
+That's it!
+With just a few lines of code, we have implemented a chatbot using the RAG pattern, combining the capabilities of the Granite LLM, Infinispan, and Quarkus.
```

Comment:
I think at some point, I would make it clear than a lot more models/document stores are available in the Quarkus LangChain4j project if you prefer to use something else.